src/transformers/models/mistral/modeling_mistral.py

```diff
@@ -1202,11 +1208,16 @@ def forward(
         )
 
         hidden_states = outputs[0]
-        logits = self.lm_head(hidden_states)
-        logits = logits.float()
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        if num_logits_to_keep is None:
+            logits = self.lm_head(hidden_states)
+        else:
+            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :])
```

Comment:
Note on BC: despite [torch's type promotion](https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc), this change can be breaking on external scripts. For instance, if a user decides to add a FP16 tensor to `logits`, previously it would accumulate in FP32 and now it can accumulate in FP16.

Because of this, I'd suggest going through a deprecation cycle, e.g.
```py
        if labels is None:
            logger.warning_once(
                'Starting from v4.44, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)'
            )
        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
        if num_logits_to_keep is None:
            logits = self.lm_head(hidden_states).float()
        else:
            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :]).float()
```

Then on v4.44 we would remove the `.float()`, giving the users time to adapt :)